### DISCUSSION TITLE
Benefit for GraphQl in monolithic project

### DISCUSSION THREAD

**Discussion by Virtualcourager** [graphql-dotnet/graphql-dotnet#3685] [General]
Does anyone use GraphQl in monolithic projects? Is there any benefit compared with REST API?

I am new to this and making decisions on whether I should try graphQl on my monolithic project (Definitely, not trying to migrate it to microservice). Personally, I think most of the benefits of GraphQl are for microservices. I am not sure whether it is also 'strongly recommended' in monolithic projects.

**Comment by Shane32** [comment]
I use it in a 'monolithic' server project which has maybe 70-80 database tables.  However, the front end SPA is stored in a separate solution, so is it really 'monolithic' at that point?

I find that the dynamic nature of GraphQL is such that it allows more flexibility of the front-end developers to write code that is not as dependent on the back-end design.  So when the dev wants to include data from a related entity in a list -- for example, the username that relates to a log entry, or a category name that relates to a product -- the dev can do so often with no changes to the back end, and still have a performant web application.  It's also very convenient for the front-end dev, as the exact shape of the result is defined in the request.

It also allows for excellent use of database indexes, since only the fields needed by the front-end (which are specified in the GraphQL request) are pulled from the database.  For the prior examples, a REST API would need to either (1) already provide the related data, even if it was not needed by the client, or (2) the client would have to make additional call(s) to retrieve the extra information, piecing it together.  Neither is an attractive option, so for best performance this often means custom REST API calls designed as dictated by the front-end dev.  Of course within GraphQL you still design based on the front-end requirements, but generally customizations are either (1) search code, or (2) mutations -- the retrieval portion is fully dynamic and mostly unrelated to the requirements of the front-end.

On the flip side of the performance benefits, it is easy to construct a GraphQL request which is extremely taxing for the database engine to perform.  Because it is so easy (good), sometimes front-end developers write code that is not performant (bad).  Whereas a REST API would have an easily-quantifiable database hit per request, which is easily understood to both the front-end and back-end devs.  For example, simply counting REST API calls might give a good approximation of database requirements.  But with GraphQL, if there's a performance issue, it is much more challenging to log/analyze.

GraphQL also has a well-defined notification system in the form of websocket-based subscriptions, which is much better (in my opinion) than webhooks, as it requires the receiver to establish and maintain the connection.  Again, the receiver can dictate the exact form of the data that is required in each notification, which allows for very efficient notifications.

Finally, I think that the benefits of GraphQL Federation, even if not in use initially, are worth mentioning for large projects.  It may be that you have only one piece, such as authentication or data storage, that is not part of the main database/application; GraphQL Federation makes it possible to link both projects together in a seamless fashion.

**Reply by Virtualcourager** [reply]
Thanks for this detailed sharing. Your points are quite insightful.
